### PR TITLE
Small improvements to kubeadm feature-gates

### LINKED ISSUE #393: Move package `app/cmd/features` to `app/features`
I don't think the generic `features` package should be under `cmd/kubeadm/app/cmd/features`, but instead `cmd/kubeadm/app/features`. Should be easy to fix.

cc @jamiehannaford could you do this?

### LINKED ISSUE #394: Add CLI flag for `cfg.FeatureFlags`
As part of adding feature flags, we should add the CLI flag extension as well so it's easy to enable/disable features.

cc @jamiehannaford you did this work; do you wanna finish this :smile:?

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR implements two small improvements for kubeadm feature-gates:
- Move the generic features package from `cmd/kubeadm/app/cmd/features`to ` cmd/kubeadm/app/features`
- Add CLI flag for `cfg.FeatureFlags`

**Which issue this PR fixes**: 
fixes [#393](https://github.com/kubernetes/kubeadm/issues/393)
fixes [#394](https://github.com/kubernetes/kubeadm/issues/394)

**Special notes for your reviewer**:
Main work of this PR is grouped into two commits, one for each issue + a separated commit for autogenerated bazel files.
cc @luxas 
cc @jamiehannaford

### CODE REVIEW

File: cmd/kubeadm/app/cmd/phases/selfhosting.go [hunk 1]
```diff
@@ -48,5 +57,8 @@ func NewCmdSelfhosting() *cobra.Command {
 	}
 
 	cmd.Flags().StringVar(&kubeConfigFile, "kubeconfig", "/etc/kubernetes/admin.conf", "The KubeConfig file to use for talking to the cluster")
+	cmd.Flags().StringVar(&featureFlagsString, "feature-gates", featureFlagsString, "A set of key=value pairs that describe feature gates for alpha/experimental features. "+
```
Comment (luxas) [inline cmd/kubeadm/app/cmd/phases/selfhosting.go]: What about a helper func that would look something like this? Not sure how fancy we wanna get, but:
```go
// in features.go
type ChosenFeatures struct {
     featureGatesString *string
}

func AddFeaturesToFlagSet(fs *pflag.FlagSet, featureList *FeatureList) *ChosenFeatures {
    cf := &ChosenFeatures{}
    fs.StringVar(cf.featureGatesString, "feature-gates", "", "A set of key=value pairs that describe feature gates for alpha/experimental features. Options are:\n"+strings.Join(features.KnownFeatures(&features.InitFeatureGates), "\n"))
    return cf
}

func (cf *ChosenFeatures) GetFeaturesEnabledMap() (map[string]bool, error) {
    // use *cf.featureGatesString here
}
```

Then usage would be pretty straightforward:
```go
var enabledFeatures features.ChosenFeatures

// in cobra Run() {
    var err error
    cfg.FeatureGates, err = enabledFeatures.GetFeaturesEnabledMap()
    if err != nil {
         return err
    }
}

enabledFeatures := AddFeaturesToFlagSet(cmd.Flags(), &features.InitFeatureGates)
```

Not sure if this is a better or worse approach. I'll leave it up to you to decide.
We can always rework it later if we find the need for it.

File: cmd/kubeadm/app/features/features.go [hunk 2]
```diff
@@ -61,3 +66,45 @@ var InitFeatureGates = FeatureList{
...
+		if len(s) == 0 {
+			continue
+		}
+		arr := strings.SplitN(s, "=", 2)
+		k := strings.TrimSpace(arr[0])
+
+		if !Supports(*f, k) {
+			return nil, fmt.Errorf("unrecognized feature-gate key: %s", k)
+		}
+		if len(arr) != 2 {
```
Comment (luxas) [inline cmd/kubeadm/app/features/features.go]: nit: move this right after the `arr` declaration

File: cmd/kubeadm/app/features/features.go [hunk 3]
```diff
@@ -61,3 +66,45 @@ var InitFeatureGates = FeatureList{
...
+			return nil, fmt.Errorf("missing bool value for feature-gate key:%s", k)
+		}
+		v := strings.TrimSpace(arr[1])
```
Comment (luxas) [inline cmd/kubeadm/app/features/features.go]: nit: declare `k` and `v` at the same time.

**Review by luxas — COMMENTED** [review]
Basically LGTM, but please fixup the nits I had there and add unit tests
Also I had an alternative approach, feel free to pick up or discard ;)

File: cmd/kubeadm/app/cmd/phases/selfhosting.go (hunk 1 above)
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/selfhosting.go]: I tested this approach, but I'm not sure that the additional struct makes the code more readable (the final number of lines in init.go or selfhosting.go is nearly the same). 

However, if we decide to rework later, the implementation was:

in features.go

```go
// FeatureGatesFlag represent a variable backing a --feature-flag
type FeatureGatesFlag struct {
	featureList        *FeatureList
	featureGatesString string
}

// AddFeatureGatesToFlagSet adds --feature-flag for a specific FeatureList to a given FlagSet
func AddFeatureGatesToFlagSet(fs *pflag.FlagSet, featureList *FeatureList) *FeatureGatesFlag {
	f := &FeatureGatesFlag{featureList: featureList}
	fs.StringVar(&f.featureGatesString, "feature-gates", f.featureGatesString, "A set of key=value pairs that describe feature gates for various features. Options are:\n"+strings.Join(KnownFeatures(f.featureList), "\n"))
	return f
}

// GetFeatureGatesMap returns a Map containing selected features
func (f *FeatureGatesFlag) GetFeatureGatesMap() (map[string]bool, error) {
	return NewFeatureGate(f.featureList, f.featureGatesString)
}
```

in init.go or selfhosting.go

```go
var featureGatesFlag *features.FeatureGatesFlag

// in cobra Run() {
   var err error
   if cfg.FeatureFlags, err = featureGatesFlag.GetFeatureGatesMap(); err != nil {
      kubeadmutil.CheckErr(err)
   }
}

featureGatesFlag = features.AddFeatureGatesToFlagSet(cmd.Flags(), &features.InitFeatureGates)
```

**Comment by luxas** [comment]
@fabriziopandini Please rebase and squash your commits to one for the directory move (along with bazel files) and one for the `--feature-gates` flag. That way we'd get the changes atomic. Thanks!

**Comment by luxas** [comment]
/lgtm

Thank you @fabriziopandini! This seems to need a rebase now though...

**Review by luxas — APPROVED** [review]
/lgtm

**Review by luxas — COMMENTED** [review]
/lgtm

**Pull request merged**